File: Matlab/falseRule.m
function falseRule(f,Xi,Xs,Tol,Iter)
format long;

Yi=f(Xi); 
Ys=f(Xs); 

if Yi==0 
    fprintf('Xi is root\n'); 
else
    if Ys==0 
        fprintf('Xs is root\n');
    else
        if Yi*Ys<0 
            Xm=(Xi)-((f(Xi)*(Xi-Xs))/(f(Xi)-f(Xs)));
            Ym=f(Xm); 
            Error=Tol+1; 
            Cont=1;
            Z=[Cont,Xi,Xs,Xm,Ym,Error];
            while Ym~=0 & Error>Tol & Cont<Iter 
            
                if Yi*Ym<0
                    Xs=Xm;
                    Ys=Ym;
                else
                    Xi=Xm;
                    Yi=Ym;
                end
                    Xaux=Xm;
                    Xm=(Xi)-((f(Xi)*(Xi-Xs))/(f(Xi)-f(Xs)));
                    Ym=f(Xm);
                    Error=abs(Xm-Xaux)/Xm;
                    Cont=Cont+1;
                    Z(Cont,1)=Cont;
                    Z(Cont,2)=Xi;
                    Z(Cont,3)=Xs;
                    Z(Cont,4)=Xm;
                    Z(Cont,5)=Ym;
                    Z(Cont,6)=Error;
            end
            if Ym==0
                fprintf('%g is a root\n\n',Xm);
            else
                if Error<Tol
                    fprintf( '%g is an approximation to a root with a tolerance %g \n\n',Xm,Tol);
                else
                    fprintf('Failure in %g iterations\n\n',Iter);
                end
            end
            fprintf('\nIterations | Xi |  Xs | Xm | Ym | Error\n');
            disp(Z);
        else
            fprintf('The interval is inadequate\n\n');
        end
    end
end


ezplot(f);
grid on

File: NumericSquadProject/numericApp/methods/Roots/falseRule.py
import sympy as sm

x = sm.symbols('x')

def falseRule(f, xi, xf, tol, niter):
    
    error = 1000
    if(float(f.subs(x,xi))*float(f.subs(x,xf)) == 0):
        if(float(f.subs(x,xi)) == 0):
            return(str(round(xi,8)) + " is a root"),[]
        else:
            return(str(round(xf,8)) + " is a root"),[]
    
    elif(float(f.subs(x,xi))*float(f.subs(x,xf)) > 0):
        return("Invalid interval"),[]
    
    else:
        fxi = float(f.subs(x,xi))
        fxf = float(f.subs(x,xf))
        pm = (fxf*xi-fxi*xf)/(fxf-fxi)
        fpm = float(f.subs(x,pm))
        matrix = []
        matrix.append([1,xi,pm,xf,fpm,error])
        # print(matrix[-1])
        iter = 2
        while(error >= tol and float(f.subs(x,pm)) != 0 and iter <= niter):
            if(fxi*fpm < 0):
                xf = pm
            else:
                xi = pm

            p0 = pm
            pm = (float(f.subs(x,xf))*xi-float(f.subs(x,xi))*xf)/(float(f.subs(x,xf))-float(f.subs(x,xi)))

            fpm = float(f.subs(x,pm))
            error = abs(pm - p0)

            matrix.append([iter,xi,pm,xf,fpm,error])
            # print(matrix[-1])
            iter = iter +1

        if(float(f.subs(x,pm)) == 0):
            return(str(round(pm,8)) + " is root and was found in the iteration " + str(iter)),matrix
        else: 
            return(str(round(pm,8)) + " is root with tolerance " + str(tol) + " and was found in the iteration " + str(iter)),matrix

# f = sm.sympify("log(sin(x)^2 + 1)-(1/2)")
# xi = 0     
# xf = 1
# tol = 1e-7
# niter = 100
# print(falseRule(f,xi,xf,tol, niter))